pkg/volume/flocker/plugin.go

```diff
@@ -0,0 +1,138 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flocker
+
+import (
+	"github.com/docker/docker/pkg/units"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	return []volume.VolumePlugin{&flockerPlugin{}}
+}
+
+type flockerPlugin struct {
+	host volume.VolumeHost
+}
+
+type flocker struct {
+	datasetName string
+	datasetID   string
+	maximumSize int64
+	pod         *api.Pod
+	mounter     mount.Interface
+	plugin      *flockerPlugin
+}
+
+func (p *flockerPlugin) Init(host volume.VolumeHost) {
+	p.host = host
+}
+
+func (p flockerPlugin) Name() string {
+	return "kubernetes.io/flocker"
+}
+
+func (p flockerPlugin) CanSupport(spec *volume.Spec) bool {
+	return (spec.PersistentVolume != nil && spec.PersistentVolume.Spec.Flocker != nil) ||
+		(spec.Volume != nil && spec.Volume.Flocker != nil)
+}
+
+func (p *flockerPlugin) getFlockerVolumeSource(spec *volume.Spec) (*api.FlockerVolumeSource, bool) {
+	// AFAIK this will always be r/w, but perhaps for the future it will be needed
+	readOnly := false
+
+	if spec.Volume != nil && spec.Volume.Flocker != nil {
+		return spec.Volume.Flocker, readOnly
+	}
+	return spec.PersistentVolume.Spec.Flocker, readOnly
+}
+
+func (p *flockerPlugin) NewBuilder(spec *volume.Spec, pod *api.Pod, opts volume.VolumeOptions) (volume.Builder, error) {
+	source, readOnly := p.getFlockerVolumeSource(spec)
+
+	var (
+		size int64
+		err  error
+	)
+	if source.Size != "" {
+		size, err = units.FromHumanSize(source.Size)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	builder := flockerBuilder{
+		flocker: &flocker{
+			datasetName: source.DatasetName,
+			datasetID:   "",
+			maximumSize: size,
+			pod:         pod,
+			mounter:     p.host.GetMounter(),
+			plugin:      p,
+		},
+		exe:      exec.New(),
+		opts:     opts,
+		readOnly: readOnly,
+	}
+	return &builder, nil
+}
+
+func (p *flockerPlugin) NewCleaner(datasetName string, podUID types.UID) (volume.Cleaner, error) {
+	return nil, nil
+}
+
+type flockerBuilder struct {
+	*flocker
+	exe      exec.Interface
+	opts     volume.VolumeOptions
+	readOnly bool
+}
+
+func (b flockerBuilder) GetPath() string {
+	return "/flocker/" + b.flocker.datasetID
+}
+
+func (b flockerBuilder) SetUp() error {
+	return b.SetUpAt(b.GetPath())
+}
+
+func (b flockerBuilder) SetUpAt(dir string) error {
+	var (
+		err error
+		c   *flockerClient
+	)
+	if b.flocker.maximumSize > 0 {
+		c, err = newFlockerClientWithSize(b.pod, b.flocker.maximumSize)
+	} else {
+		c, err = newFlockerClientWithoutSize(b.pod)
+	}
+	if err != nil {
+		return err
+	}
+
+	datasetID, err := c.CreateVolume(dir)
```

Comment:
flocker doesn't need to mount the volume? how does containers access to volume? if docker mounts the volume, is there any docker release dependency?